// app/game/index.js
var weaponsModule = require('app/weapons'),
    playersModule = require('app/players'),
    countriesScore = {};

module.exports = {
    tryGetGameInfo: function(playerName, playerKey) {
        var player = playersModule.playersHash[playerName];
        return {
            playerData: player && player.key === playerKey ? player : null,
            playersHash: playersModule.playersHash
        }
    },

    shot: function(playerName, key, targetPlayerName, selectedWeaponName) {
        var player = playersModule.playersHash[playerName],
            weapon = player && player.getWeaponByName(selectedWeaponName),
            targetPlayer = playersModule.playersHash[targetPlayerName];
        if(!player || !weapon || !targetPlayer || player.key !== key) {
            return false;
        }
        var shotResults = weapon.shot(targetPlayer, playersModule.getPlayersArray(playerName));
        shotResults.forEach(function(result) {
            if(result.isKill) {
                countriesScore[player.country.demonym] = (countriesScore[player.country.demonym] || 0) + 1;
            }
            player.actionList.push({
                type: 'shot',
                date: Date.now(),
                params: {
                    damage: result.damage,
                    isKill: result.isKill,
                    weaponName: selectedWeaponName
                }
            });
        });
        return shotResults;
    },
};
