// app/players/index.js
var weaponsModule = require('app/weapons'),
    fs = require('fs'),
    path = require('path'),
    utils = require('app/utils');

function generateGUID() {
    function s4() {
        return Math.floor((1 + Math.random()) * 0x10000)
            .toString(16)
            .substring(1);
    }

    return s4() + s4() + '-' + s4() + '-' + s4() + '-' +
        s4() + '-' + s4() + s4() + s4();
}

module.exports = {
    playersHash: {},
    bots: [],
    isPlayersExists: function(name) {
        return !!this.playersHash[name];
    },
    tryAddPlayer: function(playerData, isBot) {
        if(this.isPlayersExists(playerData.name)) {
            return {
                success: false
            }
        }
        return {
            success: true,
            playerData: this.addPlayer(playerData, isBot)
        }
    },
    addPlayer: function(palyerData, isBot) {
        if(!palyerData.country) {
            palyerData.country = this.getNoneCountry();
        }
        var player = {
            name: palyerData.name,
            key: generateGUID(),
            weaponList: weaponsModule.getWeaponList(),
            hp: 100,
            country: palyerData.country,
            town: palyerData.town,
            lastRequestTime: Date.now(),
            isBot: isBot,
            lat: palyerData.lat || palyerData.country.latlng[0] + utils.random(-5, 5),
            lng: palyerData.lon || palyerData.country.latlng[1] + utils.random(-5, 5)
        };
        this.playersHash[player.name] = player;
        if(this.bots.length === 0) {
            this.addBots();
        }
        return player;
    },
    deletePlayer: function(name) {
        delete players[name]
    },
    getNoneCountry: function(country) {
        return {
            demonym: 'NoCountry',
            latlng: [utils.random(65), utils.random(180)]
        }
    },
    addBots: function() {
        for(var i = 0; i < 5; i++) {
            this.addBot();
        }
    },
    addBot: function() {
        var me = this;
        fs.readFile(path.join(__dirname, 'randomnicknames.txt'), function(err, data) {
            var names = data.toString().split("\r\n");
            utils.getCountryList(function(countryList) {
                var bot = me.tryAddPlayer({
                    name: utils.getRandomArrayEntry(names),
                    country: utils.getRandomArrayEntry(countryList)
                }, true);
                me.bots.push(bot.playerData);
            });
        });
    },
    tryGetGameInfo: function(playerName, playerKey) {
        player = this.playersHash[playerName];
        return {
            playerData: player && player.key === playerKey ? player : null,
            playersHash: this.playersHash
        }
    }
};
