// app/players/index.js
"use strict";

var weaponsModule = require('app/weapons'),
    fs = require('fs'),
    path = require('path'),
    utils = require('app/utils');

function generateGUID() {
    function s4() {
        return Math.floor((1 + Math.random()) * 0x10000)
            .toString(16)
            .substring(1);
    }

    return s4() + s4() + '-' + s4() + '-' + s4() + '-' +
        s4() + '-' + s4() + s4() + s4();
}

class Player {
    constructor(playerData) {
        this.name = playerData.name;
        this.key = generateGUID();
        this.weaponList = weaponsModule.getWeaponList();
        this.hp = 100;
        this.lastDeathTime = 0;
        this.country = playerData.country;
        this.town = playerData.town;
        this.lastRequestTime = Date.now();
        this.isBot = playerData.isBot;
        this.lat = playerData.lat || playerData.country.latlng[0] + utils.random(-5, 5);
        this.lng = playerData.lon || playerData.country.latlng[1] + utils.random(-5, 5);
        this.actionList = [];
    }

    processDamage(damage) {
        var isKill = false;
        if(this.isImmune()) {
            damage: 0;
        } else {
            this.hp -= damage;
            if(this.hp <= 0) {
                isKill = true;
                this.lastDeathTime = Date.now();
                this.hp = 100;
            }
        }
        return {
            damage: damage,
            isKill: isKill
        }
    }
    shot(target, selectedWeaponName) {
        var weapon = this.getWeapon(selectedWeaponName);
        if(!weapon) {
            return false;
        }
        var shotResult = target.processDamage(weapon.getDamage());
        this.actionList.push({
            type: 'shot',
            date: Date.now(),
            params: {
                damage: shotResult.damage,
                isKill: shotResult.isKill,
                weaponName: selectedWeaponName
            }
        });
        return shotResult;
    }
    isImmune() {
        return Date.now() - this.lastDeathTime < 15 * 1000;
    }
    getWeapon(weaponName) {
        return this.weaponList.find(function(weapon) {
            return weapon.name === weaponName;
        });
    }
}

module.exports = {
    playersHash: {},
    bots: [],
    isPlayersExists: function(name) {
        return !!this.playersHash[name];
    },
    tryAddPlayer: function(playerData) {
        if(this.isPlayersExists(playerData.name)) {
            return {
                success: false
            }
        }
        return {
            success: true,
            playerData: this.addPlayer(playerData)
        }
    },
    addPlayer: function(playerData) {
        if(!playerData.country) {
            playerData.country = this.getNoneCountry();
        }
        var player = new Player(playerData);
        this.playersHash[player.name] = player;
        if(this.bots.length === 0) {
            this.addBots();
        }
        return player;
    },
    deletePlayer: function(name) {
        delete players[name]
    },
    getNoneCountry: function(country) {
        return {
            demonym: 'NoCountry',
            latlng: [utils.random(65), utils.random(180)]
        }
    },
    addBots: function() {
        for(var i = 0; i < 5; i++) {
            this.addBot();
        }
    },
    addBot: function() {
        var me = this;
        fs.readFile(path.join(__dirname, 'randomnicknames.txt'), function(err, data) {
            var names = data.toString().split("\r\n");
            utils.getCountryList(function(countryList) {
                var bot = me.tryAddPlayer({
                    name: utils.getRandomArrayEntry(names),
                    country: utils.getRandomArrayEntry(countryList),
                    isBot: true
                }, true);
                me.bots.push(bot.playerData);
            });
        });
    }
};
